dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/ClientRateLimitHelper.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2024 OICR and UCSC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.dockstore.webservice.helpers;
+
+import io.dockstore.webservice.CustomWebApplicationException;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import org.apache.commons.collections.map.CaseInsensitiveMap;
+import org.apache.http.HttpStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Helps a Java Client making http requests stay within the server's rate limits.
+ *
+ * Defaults to honoring the header fields from https://datatracker.ietf.org/doc/draft-ietf-httpapi-ratelimit-headers/. Note that GitHub
+ * precedes the header names with "x-", a common practice for non-standard headers (the link is a draft, not formally accepted).
+ *
+ * <ul>
+ *     <li>RateLimit-Limit</li>
+ *     <li>RateLimit-Remaining</li>
+ *     <li>RateLimit-Reset</li>
+ * </ul>
+ */
+public class ClientRateLimitHelper {
+    // public static final String LIMIT_HEADER = "X-RateLimit-Limit";
+    public static final String REMAINING_HEADER = "X-RateLimit-Remaining";
+    public static final String RESET_HEADER = "X-RateLimit-Reset";
+    /**
+     * The padding to keep from 0 remaining requests. If the &quot;X-RateLimit-Remaining&quut; header has a value
+     * smaller than this, wait for a rate limit reset before proceeding. This
+     */
+    private static final int REQUESTS_REMAINING_PADDING = 10;
```

Comment:
Would we ever want to change this value, and if so, would it make sense to pass this parameter to the constructor?

Reply:
I'm hoping we don't need to use this class elsewhere, although it's somewhat written so that it could be if needed. Yes, I suppose if the rate limit were, say, 2000 requests a minute, unlike Zenodo's 133, you might want a higher padding. I could alternatively calculate, say to 10% of the limit, but it's a bit of a guess in any case, so I'll move it into the constructor.